File: mars4/src/main/java/edu/missouristate/mars/MarsSplashScreen.java
package edu.missouristate.mars;

import javax.swing.*;
import java.awt.*;

/**
 * Produces MARS splash screen.<br>
 * Adapted from <a href="http://www.java-tips.org/content/view/1267/2/">...</a><br>
 */

public class MarsSplashScreen extends JWindow {
    private final int duration;

    public MarsSplashScreen(int d) {
        duration = d;
    }

    /**
     * A simple little method to show a title screen in the center
     * of the screen for the amount of time given in the constructor
     */
    public void showSplash() {
        ImageBackgroundPanel content = new ImageBackgroundPanel();
        this.setContentPane(content);

        // Set the window's bounds, centering the window
        // Wee bit of a hack.  I've hardcoded the image dimensions of 
        // MarsSurfacePathfinder.jpg, because obtaining them via
        // getHeight() and getWidth() is not trival -- it is possible
        // that at the time of the call the image has not completed
        // loading so the Image object doesn't know how big it is.
        // So observers are involved -- see the API.
        int width = 390;
        int height = 215;
        Toolkit tk = Toolkit.getDefaultToolkit();
        Dimension screen = tk.getScreenSize();
        int x = (screen.width - width) / 2;
        int y = (screen.height - height) / 2;
        setBounds(x, y, width, height);

        // Build the splash screen
        JLabel title = new JLabel("MARS: Mips Assembler and Runtime Simulator", JLabel.CENTER);
        JLabel copyrt1 = new JLabel("<html><br><br>Version " + Globals.version + " Copyright (c) " + Globals.copyrightYears + "</html>", JLabel.CENTER);
        JLabel copyrt2 = new JLabel("<html><br><br>" + Globals.copyrightHolders + "</html>", JLabel.CENTER);
        title.setFont(new Font("Sans-Serif", Font.BOLD, 16));
        title.setForeground(Color.black);
        copyrt1.setFont(new Font("Sans-Serif", Font.BOLD, 14));
        copyrt2.setFont(new Font("Sans-Serif", Font.BOLD, 14));
        copyrt1.setForeground(Color.white);
        copyrt2.setForeground(Color.white);

        content.add(title, BorderLayout.NORTH);
        content.add(copyrt1, BorderLayout.CENTER);
        content.add(copyrt2, BorderLayout.SOUTH);

        // Display it
        setVisible(true);
        // Wait a little while, maybe while loading resources
        try {
            Thread.sleep(duration);
        } catch (Exception ignored) {}
        setVisible(false);
    }

    @SuppressWarnings("ThrowablePrintedToSystemOut")
    static class ImageBackgroundPanel extends JPanel {
        Image image;

        public ImageBackgroundPanel() {
            try {
                image = new ImageIcon(Toolkit.getDefaultToolkit().getImage(this.getClass().getResource(Globals.imagesPath + "MarsSurfacePathfinder.jpg"))).getImage();
            } catch (Exception e) {
                System.out.println(e); /*handled in paintComponent()*/
            }
        }

        @Override
        protected void paintComponent(Graphics g) {
            super.paintComponent(g);
            if (image != null) g.drawImage(image, 0, 0, this.getWidth(), this.getHeight(), this);
        }
    }
}

File: src/main/java/edu/missouristate/mars/MarsSplashScreen.kt
package edu.missouristate.mars

import java.awt.*
import javax.swing.ImageIcon
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JWindow

/**
 * Produces MARS splash screen.<br></br>
 * Adapted from [...](http://www.java-tips.org/content/view/1267/2/)<br></br>
 */
class MarsSplashScreen(private val duration: Int) : JWindow() {
    /**
     * A simple little method to show a title screen in the center
     * of the screen for the amount of time given in the constructor
     */
    fun showSplash() {
        val content = ImageBackgroundPanel()
        this.contentPane = content

        // Set the window's bounds, centering the window
        // Wee bit of a hack.  I've hardcoded the image dimensions of 
        // MarsSurfacePathfinder.jpg, because obtaining them via
        // getHeight() and getWidth() is not trivial -- it is possible
        // that at the time of the call the image has not completed
        // loading so the Image object doesn't know how big it is.
        // So observers are involved -- see the API.
        val width = 390
        val height = 215
        val tk = Toolkit.getDefaultToolkit()
        val screen = tk.screenSize
        val x = (screen.width - width) / 2
        val y = (screen.height - height) / 2
        setBounds(x, y, width, height)

        // Build the splash screen
        val title = JLabel("MARS: Mips Assembler and Runtime Simulator", JLabel.CENTER)
        val copyrightLineOne = JLabel(
            "<html><br><br>Version " + Globals.version + " Copyright (c) " + Globals.copyrightYears + "</html>",
            JLabel.CENTER
        )
        val copyrightLineTwo = JLabel("<html><br><br>" + Globals.copyrightHolders + "</html>", JLabel.CENTER)
        title.font = Font("Sans-Serif", Font.BOLD, 16)
        title.foreground = Color.black
        copyrightLineOne.font = Font("Sans-Serif", Font.BOLD, 14)
        copyrightLineTwo.font = Font("Sans-Serif", Font.BOLD, 14)
        copyrightLineOne.foreground = Color.white
        copyrightLineTwo.foreground = Color.white

        content.add(title, BorderLayout.NORTH)
        content.add(copyrightLineOne, BorderLayout.CENTER)
        content.add(copyrightLineTwo, BorderLayout.SOUTH)

        // Display it
        isVisible = true
        // Wait a little while, maybe while loading resources
        try {
            Thread.sleep(duration.toLong())
        } catch (ignored: Exception) {
        }
        isVisible = false
    }

    internal class ImageBackgroundPanel : JPanel() {
        private var image: Image? = null

        init {
            try {
                image = ImageIcon(
                    Toolkit.getDefaultToolkit()
                        .getImage(this.javaClass.getResource(Globals.imagesPath + "MarsSurfacePathfinder.jpg"))
                ).image
            } catch (e: Exception) {
                println(e) /*handled in paintComponent()*/
            }
        }

        override fun paintComponent(g: Graphics) {
            super.paintComponent(g)
            if (image != null) g.drawImage(image, 0, 0, this.width, this.height, this)
        }
    }
}
